### ISSUE TITLE
Why the predictions of TabM are averaged only on inference, but not during training?

### ISSUE THREAD

**Issue by isCopyman** [yandex-research/tabm#6]
## Description

**Issue Summary:**

In both the paper and the provided code, TabM and TabM-mini compute the output for each component individually. They then replicate the true labels (`true_y`) to match the number of components, calculate the loss for each component, perform backpropagation, and update the parameters. Mathematically, the loss is computed as:
$`\mathcal{L}_1 = \frac{1}{k} \sum_{i=1}^{k} (y_i - y)^2`$

However, if we slightly modify the workflow for a modified version of TabM, where in the `forward` method of the `Model` class, we first average all the outputs and then compute the loss with a single `true_y`, the loss would be:

$`\mathcal{L}_2 = \left(\frac{1}{k} \sum_{i=1}^{k} y_i - y\right)^2`$

Questions:
1. **Difference in Behavior:**
What differences arise from these two loss computation methods?
2. **Performance Impact:**
Would there be any performance differences between the two approaches?
3. **Experimental Validation:**
Have you conducted any experiments related to this modification?

**Additional Details:**

- **Mathematical Implications:**  

  From a mathematical standpoint, it seems that the gradient propagation would differ between the two loss functions.

- **Code Modification:**  

  The code modification is straightforward. You only need to add the following line before the `return` statement in the `forward` method of the `Model` class:
``` 
x = x.mean(dim=1, keepdim=True)
````
**Example of Modified `forward` Method:**
``` python
def forward(
    self, x_num: None | Tensor = None, x_cat: None | Tensor = None
) -> Tensor:
    x = []
    if x_num is not None:
        x.append(x_num if self.num_module is None else self.num_module(x_num))
    if x_cat is None:
        assert self.cat_module is None
    else:
        assert self.cat_module is not None
        x.append(self.cat_module(x_cat).float())
    x = torch.column_stack([x_.flatten(1, -1) for x_ in x])

    if self.k is not None:
        x = x[:, None].expand(-1, self.k, -1)
        if self.minimal_ensemble_adapter is not None:
            x = self.minimal_ensemble_adapter(x)
    else:
        assert self.minimal_ensemble_adapter is None

    x = self.backbone(x)
    x = self.output(x)



    if self.k is None:
        x = x[:, None]
    ####################Add this##############################
    # Modified line to average the outputs before loss calculation
    x = x.mean(dim=1, keepdim=True)
    ###############################################################
    return x
```

**Comment by isCopyman** [comment]
I tried it and there was a noticeable drop in performance compared to the original tabM.
original:
![image](https://github.com/user-attachments/assets/0bedab52-6464-4998-a151-35920ba6fcdb)

modified:
![image](https://github.com/user-attachments/assets/2adcc2eb-2324-4b43-8be8-cfbe47ce4c50)

**Comment by Yura52** [comment]
**The current design is intentional: the $k$ predictions of TabM must be trained separately.** So it is critical _not_ to average predictions when _training_ TabM, and do that _only on inference_.

The paper ["Joint Training of Deep Ensembles Fails Due to Learner Collusion"](https://arxiv.org/abs/2301.11323) explains why it is criticial to train ensemble members separately, without averaging their predictions. Thus, our design is not specific to TabM, it is a common implementation detail for any ensemble-like system. In particular, it is expected that averaging predictions leads to bad results, as confirmed by the above experiments.
